src/devices/src/virtio/net/iovec.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use crate::virtio::DescriptorChain;
+use std::io::IoSlice;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use vm_memory::{GuestMemory, GuestMemoryError, GuestMemoryMmap};
+
+#[derive(Debug)]
+pub enum Error {
+    /// We found a descriptor with unexpected access rights
+    DescriptorAccessMode,
+    /// An error happened with guest memory handling
+    GuestMemory(vm_memory::GuestMemoryError),
+}
+
+impl From<GuestMemoryError> for Error {
+    fn from(err: GuestMemoryError) -> Self {
+        Error::GuestMemory(err)
+    }
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub(crate) struct IoVec<'a> {
+    // container of the memory regions included in this IO vector
+    vecs: Vec<IoSlice<'a>>,
+    // IoVecs can contain memory regions that are either read-only
+    // or write-only
+    is_write_only: bool,
+    // Total length of the IoVec
+    len: usize,
+    // marker
+    phantom: PhantomData<DescriptorChain<'a>>,
+}
+
+impl<'a> Deref for IoVec<'a> {
+    type Target = [IoSlice<'a>];
+
+    fn deref(&self) -> &Self::Target {
+        self.vecs.as_slice()
+    }
+}
+
+impl<'a> IoVec<'a> {
+    /// Create an `IoVec` from a `DescriptorChain`
+    pub fn from_descriptor_chain(
+        mem: &'a GuestMemoryMmap,
+        head: DescriptorChain,
+        is_write_only: bool,
+    ) -> Result<Self> {
+        let mut iovec = Self {
+            vecs: vec![],
+            is_write_only,
+            len: 0,
+            phantom: PhantomData,
+        };
+
+        let mut next_descriptor = Some(head);
+
+        while let Some(desc) = next_descriptor {
+            if desc.is_write_only() != is_write_only {
+                return Err(Error::DescriptorAccessMode);
+            }
+
+            let ptr = mem.get_host_address(desc.addr)?;
+            let slice = unsafe { std::slice::from_raw_parts(ptr, desc.len as usize) };
+            iovec.vecs.push(IoSlice::new(slice));
+            iovec.len += desc.len as usize;
+
+            next_descriptor = desc.next_descriptor();
+        }
+
+        Ok(iovec)
+    }
+
+    /// Get the total length of the memory regions covered by this `IoVec`
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    // This assumes that `IoVec` has `size` bytes available to copy from
+    fn read_bytes_unchecked(&self, start: usize, size: usize) -> Vec<u8> {
```

Comment:
I do like this approach. In fact, I was already thinking to have the caller passing us the buffer to which we should write the result, so that we can re-use buffers and avoid creating `Vec`s all the time.